aten/src/ATen/native/SoftMax.cpp

```diff
@@ -115,6 +115,62 @@ void host_softmax_backward(
         }
       });
 }
+
+template <typename scalar_t>
+void host_masked_softmax(Tensor output, const Tensor& input, const Tensor& mask, const int64_t dim) {
+  int64_t outer_size = 1;
+  int64_t dim_size = input.size(dim);
+  int64_t inner_size = 1;
+  for (int64_t i = 0; i < dim; ++i)
+    outer_size *= input.size(i);
+  for (int64_t i = dim + 1; i < input.dim(); ++i)
+    inner_size *= input.size(i);
+  int64_t dim_stride = inner_size;
+  int64_t outer_stride = dim_size * dim_stride;
+  bool* mask_data_base = mask.data_ptr<bool>();
+  scalar_t* input_data_base = input.data_ptr<scalar_t>();
+  scalar_t* output_data_base = output.data_ptr<scalar_t>();
+  int64_t grain_size = std::min(internal::GRAIN_SIZE / dim_size, (int64_t)1);
+  parallel_for(
+      0, outer_size * inner_size, grain_size,
+      [&](int64_t begin, int64_t end) {
+        for (int64_t i = begin; i < end; i++) {
+          int64_t outer_idx = i / inner_size;
+          int64_t inner_idx = i % inner_size;
+          bool* mask_data =
+              mask_data_base + outer_idx * outer_stride + inner_idx;
+          scalar_t* input_data =
+              input_data_base + outer_idx * outer_stride + inner_idx;
+          scalar_t* output_data =
+              output_data_base + outer_idx * outer_stride + inner_idx;
+          scalar_t max_input = input_data[0];
+
+          std::vector<int64_t> unmasked;
+          unmasked.reserve(dim_size);
+
+          for (int64_t d = 0; d < dim_size; d++){
+            if (mask_data[d * dim_stride] == false) {
+              max_input = std::max(max_input, input_data[d * dim_stride]);
+              unmasked.push_back(d);
+            }
+            else {
+              output_data[d * dim_stride] = 0;
+            }
+          }
+
+          acc_type<scalar_t, false> tmpsum = 0;
+          for (auto d: unmasked) {
```

Comment:
Did you find a speedup in using this additional std::vector? Otherwise I'd expect repeating the check above and iterating over the entire data again to be faster instead of saving out the indices into a freshly allocated std::vector and then iterating over those simply because whatever portion of "mask_data" you've just used is still available in cache anyway. So you could try 
```
for (int64_t d = 0; d < dim_size; d++){
    if (mask_data[d * dim_stride] == false) {
```
again instead of
```
for (auto d: unmasked) {
```